{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import seaborn as sns\n",
    "plt.style.use('dark_background')\n",
    "sns.set()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def nucl_lists():\n",
    "    \"\"\"\n",
    "    Generate lists of possible nucleotides\n",
    "    \n",
    "    Outputs:\n",
    "    di_list - list of possible dinuclotides\n",
    "    tri_list - list of possible trinuclotides\n",
    "    tetra_list - list of possible tetranuclotides\n",
    "    \"\"\"\n",
    "    \n",
    "    nucl = [\"A\",\"T\",\"C\",\"G\"]\n",
    "    di_list = [i+j for i in nucl for j in nucl]\n",
    "    tri_list = [i+j for i in di_list for j in nucl]\n",
    "    tetra_list = [i+j for i in tri_list for j in nucl]\n",
    "\n",
    "    return (di_list, tri_list, tetra_list)\n",
    "  \n",
    "    \n",
    "def rev_comp(string):\n",
    "    \"\"\"\n",
    "    Return reverse complement of DNA strand\n",
    "    \n",
    "    Input - sequence\n",
    "    \n",
    "    Output - reverse complement\n",
    "    \"\"\"\n",
    "    \n",
    "    complementary = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'} \n",
    "    comp = ''\n",
    "    for i in string:\n",
    "        comp += complementary[i]\n",
    "\n",
    "    return comp[::-1]\n",
    "\n",
    "    \n",
    "def count_occurrences(subsequence, sequence):\n",
    "    \"\"\"\n",
    "    Count occurrences of substring in a string\n",
    "    (including overlaping substrings)\n",
    "    \n",
    "    Inputs:\n",
    "    sequence - sequence\n",
    "    subsequence - subsequence to be found in sequence\n",
    "    \n",
    "    Output: number of occurrences of subsequence in sequence\n",
    "    \"\"\"\n",
    "    number_of_occurrences = 0\n",
    "    for i in range(len(sequence)):\n",
    "        if sequence[i:i+len(subsequence)] == subsequence:\n",
    "            number_of_occurrences += 1\n",
    "            \n",
    "    return number_of_occurrences\n",
    "\n",
    "\n",
    "def get_frequencies(sequence, tri_list, di_list):\n",
    "    \"\"\"\n",
    "    Count occurrences of oligonuclotides in sequence\n",
    "    \n",
    "    Input - sequence\n",
    "    \n",
    "    Outputs:\n",
    "    tetra_freq - array of tetranuclotides counts\n",
    "    tri_freq - array of trinuclotides counts\n",
    "    di_freq - array of dinuclotides counts\n",
    "    \"\"\"\n",
    "    encoding = {'A':0, 'T':1, 'C':2, 'G':3}\n",
    "    \n",
    "    seq = np.zeros(4)\n",
    "    tetra_freq = np.ones(256)\n",
    "    \n",
    "    n = 1\n",
    "    for i in sequence:\n",
    "\n",
    "        seq[0] = seq[1]\n",
    "        seq[1] = seq[2]\n",
    "        seq[2] = seq[3]\n",
    "        seq = seq*4\n",
    "        seq[3] = encoding[i]\n",
    "        \n",
    "        if n > 3:\n",
    "            tetra_freq[int(seq.sum())] += 1\n",
    "        n +=1\n",
    "    \n",
    "    #Calculate 3 and 2-mers based on 4-mers\n",
    "    di_nucl = {di_list[i]:0 for i in range(len(di_list))}\n",
    "    tri_nucl = {tri_list[i]:0 for i in range(len(tri_list))}\n",
    "    \n",
    "    for i in range(len(tetra_list)):\n",
    "        tri_nucl[tetra_list[i][0:3]] += tetra_freq[i]\n",
    "    \n",
    "    for i in tri_nucl:\n",
    "        di_nucl[i[0:2]] += tri_nucl[i]\n",
    "    \n",
    "    return(tetra_freq, tri_nucl, di_nucl)\n",
    "\n",
    "\n",
    "def rev_comp_freq(oligo_list, oligo_freq):\n",
    "    \"\"\"\n",
    "    Calculate frequencies of oligonuclotides from the complementary strand\n",
    "    \n",
    "    Inputs:\n",
    "    oligo_list - list of oligonuclotides\n",
    "    oligo_freq - list of oligonuclotides counts\n",
    "    \n",
    "    Output - updated frequencies\n",
    "    \"\"\"\n",
    "    \n",
    "    r_dir = {oligo_list[i]:oligo_freq[i] for i in range(len(oligo_list))}\n",
    "    \n",
    "    for i in range(len(oligo_list)):\n",
    "        oligo_freq[i] += r_dir[rev_comp(oligo_list[i])]\n",
    "    \n",
    "    return oligo_freq\n",
    "\n",
    "\n",
    "def exp_tnf(tetra_list, tri_nucl, di_nucl):\n",
    "    \"\"\"\n",
    "    Calculate expected tetranucleotide frequency\n",
    "    \n",
    "    Inputs:\n",
    "    tetra_list - list of tetranuclotides\n",
    "    tri_nucl - dictionary of trinuclotides frequency\n",
    "    di_nucl - dictionary of dinuclotides frequency\n",
    "    \n",
    "    Output: expected tetranuclotide frequency\n",
    "    \"\"\"\n",
    "    \n",
    "    expected_tnf = np.zeros(len(tetra_list))\n",
    "    for i in range(len(tetra_list)):\n",
    "        tn = tetra_list[i]\n",
    "        expected_tnf[i] = (tri_nucl[tn[0:3]]*tri_nucl[tn[1:4]])/di_nucl[tn[1:3]]\n",
    "            \n",
    "    return expected_tnf\n",
    "\n",
    "\n",
    "def app_variance(expected_tnf, tetra_list, tri_nucl, di_nucl):\n",
    "    \"\"\"\n",
    "    Approximated variance of tetranuclotide frequency\n",
    "    \n",
    "    Inputs:\n",
    "    expected_tnf - expected tetranuclotide frequency\n",
    "    tetra_list - list of tetranuclotides\n",
    "    tri_nucl - dictionary of trinuclotides frequency\n",
    "    di_nucl - dictionary of dinuclotides frequency\n",
    "    \n",
    "    Output - approximated variance of tetranuclotide frequency\n",
    "    \"\"\"\n",
    "    \n",
    "    app_var = np.zeros(len(tetra_list))\n",
    "    for i in range(len(tetra_list)):\n",
    "        tn = tetra_list[i]\n",
    "        n23 = di_nucl[tn[1:3]]\n",
    "        n123 = tri_nucl[tn[0:3]]\n",
    "        n234 = tri_nucl[tn[1:4]]\n",
    "        app_var[i] = expected_tnf[i] * (((n23-n123)*(n23-n234))/(n23*n23))\n",
    "            \n",
    "    return app_var\n",
    "\n",
    "\n",
    "def tnf_z_score(tetra_freq, expected_tnf, app_var):\n",
    "    \"\"\"\n",
    "    Calculate tetranuclotide z-score\n",
    "    \n",
    "    Inputs:\n",
    "    tetra_freq - observed tetranuclotide frequency\n",
    "    expected_tnf - expected tetranuclotide frequency\n",
    "    app_var - approximated variance of tetranuclotide frequency\n",
    "    \n",
    "    Output - tetranuclotide z-score\n",
    "    \"\"\"\n",
    "    \n",
    "    z_scores = (tetra_freq-expected_tnf)/np.sqrt(app_var)\n",
    "    return z_scores\n",
    "\n",
    "\n",
    "def get_z_scores(sequence, di_list, tri_list, tetra_list):\n",
    "    \"\"\"\n",
    "    Calculate z-scores from sequence\n",
    "    \n",
    "    Inputs:\n",
    "    sequence - DNA sequence\n",
    "    di_list - list of possible dinuclotides\n",
    "    tri_list - list of possible trinuclotides\n",
    "    tetra_list - list of possible tetranuclotides\n",
    "    \n",
    "    Output - np array of z-scores\n",
    "    \"\"\"\n",
    "    \n",
    "    tetra_freq, tri_nucl, di_nucl = get_frequencies(sequence, tri_list, di_list)\n",
    "    tetra_freq = rev_comp_freq(tetra_list, tetra_freq)\n",
    "    \n",
    "    expected_tnf = exp_tnf(tetra_list, tri_nucl, di_nucl)\n",
    "    app_var = app_variance(expected_tnf, tetra_list, tri_nucl, di_nucl)\n",
    "    z_scores = tnf_z_score(tetra_freq, expected_tnf, app_var)\n",
    "    \n",
    "    return z_scores\n",
    "\n",
    "\n",
    "def inter_correlations(all_z_scores):\n",
    "    \"\"\"\n",
    "    Calcualte Person's correlation coefficient between fragment and median\n",
    "    \n",
    "    Input - np array of z-scores for all fragments\n",
    "    \n",
    "    Output - np array of correlation coeeficients\n",
    "    \"\"\"\n",
    "    \n",
    "    median_z = np.median(all_z_scores, axis=0)\n",
    "\n",
    "    correlations = np.zeros(len(all_z_scores))\n",
    "    for i in range(len(all_z_scores)):\n",
    "        correlations[i] = np.corrcoef(all_z_scores[i], median_z)[1,0]\n",
    "        \n",
    "    return correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tetranucleotide Z-score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "contigs = SeqIO.parse(\"./test/bin.2.fa\", 'fasta')\n",
    "di_list, tri_list, tetra_list = nucl_lists()\n",
    "\n",
    "contigs_z_scores = []\n",
    "for record in contigs:  \n",
    "    z_scores = get_z_scores(record.seq, di_list, tri_list, tetra_list)\n",
    "    contigs_z_scores.append(z_scores)\n",
    "    \n",
    "contigs_z_scores = np.array(contigs_z_scores)\n",
    "contig_corr = inter_correlations(contigs_z_scores)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f3dbf90c550>"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXkAAAEBCAYAAACdctWRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAEtdJREFUeJzt3W1sW3fZx/GfU2fJsqQ3W7DbouUOYoAIE5RpFJYJEhVBVpJ4ZS0S6YCBykMrIJOqaVB1YZVAqN2DqEBTxTqhTUPjRTdaFqoum7bRSFOG2nuIVa2KOrGmGw9N3GS0Sbw4sf2/X6B5a+l6jo99bOfq9/OmdXx8+bqa5Gf3+Jz/iTjnnAAAJtVUugEAQHgIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMOilXriN96YUS5XmgUwm5sbNTExXZJa1YoZ7bgU5rwUZpTKO2dNTURXXnlFwY+rWMjncq5kIf9WPeuY0Y5LYc5LYUap+udkdw0AGEbIA4BhhDwAGEbIA4BhhDwAGEbIA4BhhDwAGFax4+QBeMvkpPR8JpTadbVRRXmbZx4hD1Sx9HxGh46NhVJ7RdsSReuIAOt4HQcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAwwh5ADCMkAcAw3yH/D333KPNmzdLko4dO6Y1a9bopptu0l133aVMJhNagwCA4HyF/Isvvqi9e/fmb9955526++679fTTT8s5p927d4fWIAAgOM+Q//e//60dO3Zo48aNkqR//OMfmp2d1Sc+8QlJ0po1azQ0NBRulwCAQDxD/u6779amTZu0ePFiSdL4+LhisVj+/lgsprGxsfA6BAAEFr3YnY8//riWLVum9vZ27dmzR5KUy+UUiUTy2zjnzrntV3NzY8GPuZhYrKmk9aoRM9rhd043mVJTY30oPTQ01Cl2VUMotSW+l9XioiG/f/9+JZNJrV69WmfOnFEqlVIkElEymcxvc/r0acXj8YKfeGJiWrmcK7zjC4jFmpRMTpWkVrViRjsKmTOVzmhqejaUPlKptJLZbCi1+V6WXk1NJNCb44uG/MMPP5z/+549e3Tw4EFt27ZNvb29eumll3T99dfrySefVEdHR+EdAwBCd9GQfzf333+/BgYGND09rWuvvVa33XZbqfsCAJSA75Bfs2aN1qxZI0n6yEc+oieeeCK0pgAApcEZrwBgGCEPAIYR8gBgGCEPAIYR8gBgGCEPAIYR8gBgGCEPAIYFOuMVwH9kclJ6vrCL5rjJlFJpf48p0fJOuIQR8kAR0vMZHTpW2FLbTY31vhcdW/7hmPdGwEWwuwYADCPkAcAwQh4ADGOfPMwL8uGoX3wwimpHyMO8IB+O+sUHo6h27K4BAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMNYuwa4REVqIprxeYWqQtWn5kKpi8IR8sAlKj2f1cvHk6HU7rz+fxUJpTIKxe4aADCMkAcAw9hdg4p766IebjKlVAj7iLmwBy5lhDwq7q2LejQ11mtqerbk9bmwBy5l7K4BAMMIeQAwjJAHAMN8hfwvfvELdXd3q6enRw8//LAkaWRkRIlEQl1dXdqxY0eoTQIAgvH84PXgwYP605/+pMHBQWUyGXV3d6u9vV1btmzRb37zGy1btkwbNmzQ8PCwOjs7y9EzAMAnz3fyn/rUp/Too48qGo1qYmJC2WxWZ8+eVWtrq1paWhSNRpVIJDQ0NFSOfgEABfC1u6a2tla//OUv1dPTo/b2do2PjysWe/uwtHg8rrGxsdCaBAAE4/s4+dtvv13f+c53tHHjRo2OjioSeXtlCufcObf9aG5uLGh7L7FYU0nrVSOrM7rJlJoa6yUp/2cp1dZGQ6lbTG2/j6nG3v2y+vN6vmqf0zPk//a3v2lubk5tbW26/PLL1dXVpaGhIS1atCi/TTKZVDweL+iJJyamlSvRqYixWJOSyamS1KpWlmdMpTOamp4N7WSo+flMKHWD1i5kzmrrvRBWf17fqZy/lzU1kUBvjj131/z973/XwMCA5ubmNDc3p+eee059fX06ceKETp48qWw2q3379qmjoyNQ4wCA8Hi+k+/s7NThw4f1pS99SYsWLVJXV5d6enp01VVXqb+/X+l0Wp2dnVq1alU5+gUAFMDXPvn+/n719/ef87X29nYNDg6G0hQAoDQ44xUADCPkAcAwQh4ADCPkAcAwLhoCoOQy2ZzmQrjK11vqaqOK8hbVF0IeQMml57P6v2PhLXWyom2JonXElx+8FgKAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABhGyAOAYYQ8ABjmK+QfeOAB9fT0qKenR/fee68kaWRkRIlEQl1dXdqxY0eoTQIAgvEM+ZGREb3wwgvau3evfv/73+vo0aPat2+ftmzZop07d2r//v06cuSIhoeHy9EvAKAAniEfi8W0efNmXXbZZaqtrdU111yj0dFRtba2qqWlRdFoVIlEQkNDQ+XoFwBQgKjXBh/60Ifyfx8dHdVTTz2lr33ta4rFYvmvx+NxjY2NFfTEzc2NBW3vJRZrKmm9amR1RjeZUlNjvSTl/yyl2tpoKHWLqe33MdXYu19h1m5oqFPsqobQ6hei2n8vPUP+La+88oo2bNigH/7wh1q0aJFGR0fz9znnFIlECnriiYlp5XKuoMe8m1isScnkVElqVSvLM6bSGU1Nz6qpsV5T07Mlrz8/nwmlbtDahcxZbb0XIszaqVRayWw2tPp+lfP3sqYmEujNsa8PXl966SV985vf1B133KFbbrlFS5cuVTKZzN+fTCYVj8cLfnIAQLg8Q/5f//qXvv/97+v+++9XT0+PJGn58uU6ceKETp48qWw2q3379qmjoyP0ZgEAhfHcXfPrX/9a6XRa27dvz3+tr69P27dvV39/v9LptDo7O7Vq1apQGwUAFM4z5AcGBjQwMHDB+wYHB0veEKpTJiel5zOh1C7RRzMALsD3B6+4tKXnMzp0rLAjqPxa/uGY90YAAmFZAwAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwLFrpBlA6mZyUns+EUjvnQikLIGS+Q356elp9fX361a9+pauvvlojIyPatm2b0um0vvjFL2rTpk1h9gkf0vMZHTo2Fkrt5R+OhVIXQLh87a55+eWXtW7dOo2OjkqSZmdntWXLFu3cuVP79+/XkSNHNDw8HGafAIAAfIX87t27tXXrVsXjcUnS4cOH1draqpaWFkWjUSUSCQ0NDYXaKACgcL521/zsZz875/b4+Lhisbf/+x6PxzU2VthugubmxoK29xKLNZW0XjXymtFNptTUWB/Kc9fWRstSO4znKFfvhfD7mGrs3a8wazc01Cl2VUNo9QtR7dkT6IPXXC6nSCSSv+2cO+e2HxMT08qV6NO8WKxJyeRUSWpVKz8zptIZTU3PhvL88/Ph125qrA/lOcrReyEKmbPaei9EmLVTqbSS2Wxo9f0qZ/bU1EQCvTkOdAjl0qVLlUwm87eTyWR+Vw4AoHoECvnly5frxIkTOnnypLLZrPbt26eOjo5S9wYAKFKg3TV1dXXavn27+vv7lU6n1dnZqVWrVpW6NwBAkQoK+eeffz7/9/b2dg0ODpa8IQDwEqmJaCYdzol/dbVRRQ2tBcAZrwAWnPR8Vi8fT3pvGMCKtiWK1tmJRkOvVwCA8xHyAGAYIQ8AhhHyAGAYIQ8AhhHyAGCYneOEFoBiLurhJlNKeRwXzIU9AJyPkC+jYi7q4WdRKy7sAeB87K4BAMMIeQAwjJAHAMMIeQAwjJAHAMMIeQAwjJAHAMM4Tv48xZyw5IWTlQCUGyF/nmJOWPLCyUoAyo3dNQBgGCEPAIYR8gBgGPvkAeAdIjURzXis+PoWP6vDnq+uNqpoGd9eE/IA8A7p+axePp70ta2f1WHPt6JtiaJ15YtedtcAgGGEPAAYRsgDgGGEPAAYRsgDgGGEPAAYtuAOobzQAmJBjlV9NywiBsCSBRfyF1pALMixqu+GRcQAWMLuGgAwjJAHAMMIeQAwjJAHAMOKCvk//OEP6u7uVldXlx577LFS9QQAKJHAR9eMjY1px44d2rNnjy677DL19fXp05/+tD74wQ+Wsj8AQBECh/zIyIhuuOEGvec975Ek3XTTTRoaGtIPfvADX4+vqYkEet7ooho11Nee87XL66LKZmrf5RHF1y+VYmr7mbFae/dbu5TfxwvVD0OQ2oXMWW29+68dCa32f+pXx79LkJ/Z6KKaQPkXODMDPUrS+Pi4YrG3jymPx+M6fPiw78dfeeUVQZ9aVy/7n8CP9eMDV1+5IGuHXX+h1g67Pr1fWMuSxaHVlhbuv0u5Bd4nn8vlFIm8/crinDvnNgCg8gKH/NKlS5VMvn31lGQyqXg8XpKmAAClETjkb7zxRr344ouanJzUm2++qWeeeUYdHR2l7A0AUKTA++SXLFmiTZs26bbbbtP8/Ly+/OUv6+Mf/3gpewMAFCninGPdRQAwijNeAcAwQh4ADCPkAcAwQh4ADFswIe93MbQDBw7oc5/7XBk7Ky2vOR944AGtXLlSq1ev1urVqxfkwnBeM7766qv6+te/rptvvlnf+ta3dObMmQp0WbyLzXns2LH893D16tX67Gc/q97e3gp1GpzX9/Lo0aNau3atbr75Zm3YsEFnz56tQJfF8ZpxeHhYiURCiURCd9xxh2ZmZirQ5UW4BeDUqVNu5cqV7o033nAzMzMukUi4V1555b+2SyaTbtWqVW7lypUV6LJ4fubcsGGD+/Of/1yhDovnNWMul3NdXV1ueHjYOefcfffd5+69995KtRuY359Z55xLpVKup6fHHTp0qMxdFsfPjOvWrXMHDhxwzjm3bds29/Of/7wSrQbmNeOZM2fcDTfckP/arl273E9/+tNKtXtBC+Kd/DsXQ2toaMgvhna+gYEB3wukVSM/cx45ckQPPvigEomEfvKTnyidTleo22C8Zjx69KgaGhryJ9Zt3LhRX/3qVyvVbmB+f2Yl6cEHH9SKFSv0yU9+ssxdFsfPjLlcLv/O9s0331R9fX0lWg3Ma8bR0VG9733vy6++u3LlSj377LOVaveCFkTIX2gxtLGxcy/m/eijj+qjH/2oli9fXu72SsZrzpmZGbW1tenOO+/U3r17dfbsWe3cubMSrQbmNeNrr72m9773vdqyZYtuueUWbd26VQ0NDZVotSh+fmYlaWpqSrt3716Qb078zLh582YNDAzoM5/5jEZGRtTX11fuNoviNeP73/9+nTp1Sn/9618lSU899ZROnz5d9j4vZkGEvNdiaMePH9czzzyj733ve5Vor2S85rziiiv00EMP6ZprrlE0GtX69es1PDxciVYD85oxk8no4MGDWrdunfbu3auWlhZt3769Eq0Wxe8CfoODg/r85z+v5ubmcrZXEl4zzs7O6q677tIjjzyiF154Qbfeeqt+9KMfVaLVwLxmXLx4se655x79+Mc/1tq1axWPx1VbG94Sy0EsiJD3WgxtaGhIyWRSa9eu1Xe/+12Nj4/r1ltvrUSrRfGa85///KeeeOKJ/G3nnKLRwCtTVITXjLFYTK2trfrYxz4mSert7S1oCetq4XcBv2effVbd3d3lbK1kvGY8fvy46urq8sudfOUrX9HBgwfL3mcxvGbMZrNaunSpHn/8cf3ud79TW1ubWlpaKtHqu1oQIe+1GNrtt9+up59+Wk8++aR27dqleDyu3/72txXsOBivOevr63Xffffp9ddfl3NOjz32mL7whS9UsOPCec143XXXaXJyMv/f3+eff17XXnttpdoNzM8Cfs45HT16VNddd12FuiyO14ytra06deqUXn31VUnSc889l3/xXii8ZoxEIlq/fr3GxsbknNMjjzxSfS/alfrEt1CDg4Oup6fHdXV1uV27djnnnPv2t7/tDh8+fM52r7/++oI9usY57zmHhoby92/evNml0+lKthuI14x/+ctf3Nq1a113d7dbv369O336dCXbDcxrztOnT7sbb7yxki0WzWvGAwcOuEQi4Xp7e903vvEN99prr1Wy3UC8ZvzjH//oent7XVdXl9u6daubm5urZLv/hQXKAMCwBbG7BgAQDCEPAIYR8gBgGCEPAIYR8gBgGCEPAIYR8gBgGCEPAIb9P6pdDGcgiL3fAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(contig_corr, kde=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
