{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "plt.style.use('dark_background')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def nucl_lists():\n",
    "    \"\"\"\n",
    "    Generate lists of possible nucleotides\n",
    "    \n",
    "    Outputs:\n",
    "    di_list - list of possible dinuclotides\n",
    "    tri_list - list of possible trinuclotides\n",
    "    tetra_list - list of possible tetranuclotides\n",
    "    \"\"\"\n",
    "    \n",
    "    nucl = [\"A\",\"T\",\"C\",\"G\"]\n",
    "    di_list = [i+j for i in nucl for j in nucl]\n",
    "    tri_list = [i+j for i in di_list for j in nucl]\n",
    "    tetra_list = [i+j for i in tri_list for j in nucl]\n",
    "\n",
    "    return (di_list, tri_list, tetra_list)\n",
    "  \n",
    "    \n",
    "def rev_comp(string):\n",
    "    \"\"\"\n",
    "    Return reverse complement of DNA strand\n",
    "    \n",
    "    Input - sequence\n",
    "    \n",
    "    Output - reverse complement\n",
    "    \"\"\"\n",
    "    \n",
    "    complementary = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'} \n",
    "    comp = ''\n",
    "    for i in string:\n",
    "        comp += complementary[i]\n",
    "\n",
    "    return comp[::-1]\n",
    "\n",
    "    \n",
    "def count_occurrences(subsequence, sequence):\n",
    "    \"\"\"\n",
    "    Count occurrences of substring in a string\n",
    "    (including overlaping substrings)\n",
    "    \n",
    "    Inputs:\n",
    "    sequence - sequence\n",
    "    subsequence - subsequence to be found in sequence\n",
    "    \n",
    "    Output: number of occurrences of subsequence in sequence\n",
    "    \"\"\"\n",
    "    number_of_occurrences = 0\n",
    "    for i in range(len(sequence)):\n",
    "        if sequence[i:i+len(subsequence)] == subsequence:\n",
    "            number_of_occurrences += 1\n",
    "            \n",
    "    return number_of_occurrences\n",
    "\n",
    "\n",
    "def get_frequencies(sequence):\n",
    "    \"\"\"\n",
    "    Count occurrences of oligonuclotides in sequence\n",
    "    \n",
    "    Input - sequence\n",
    "    \n",
    "    Outputs:\n",
    "    tetra_freq - array of tetranuclotides counts\n",
    "    tri_freq - array of trinuclotides counts\n",
    "    di_freq - array of dinuclotides counts\n",
    "    \"\"\"\n",
    "    encoding = {'A':0, 'T':1, 'C':2, 'G':3}\n",
    "    \n",
    "    seq = np.zeros(4)\n",
    "    tetra_freq = np.zeros(256)\n",
    "    tri_freq = np.zeros(64)\n",
    "    di_freq = np.zeros(16)\n",
    "    \n",
    "    n = 1\n",
    "    for i in sequence:\n",
    "\n",
    "        seq[0] = seq[1]\n",
    "        seq[1] = seq[2]\n",
    "        seq[2] = seq[3]\n",
    "        seq = seq*4\n",
    "        seq[3] = encoding[i]\n",
    "        \n",
    "        if n > 1:\n",
    "            di_freq[int(seq[2:5].sum())] += 1\n",
    "            \n",
    "        if n > 2:\n",
    "            tri_freq[int(seq[1:5].sum())] += 1\n",
    "            \n",
    "        if n > 3:\n",
    "            tetra_freq[int(seq.sum())] += 1\n",
    "        n +=1\n",
    "    \n",
    "    return(tetra_freq, tri_freq, di_freq)\n",
    "\n",
    "\n",
    "def rev_comp_freq(oligo_list, oligo_freq):\n",
    "    \"\"\"\n",
    "    Calculate frequencies of oligonuclotides from the complementary strand\n",
    "    \n",
    "    Inputs:\n",
    "    oligo_list - list of oligonuclotides\n",
    "    oligo_freq - list of oligonuclotides counts\n",
    "    \n",
    "    Output - updated frequencies\n",
    "    \"\"\"\n",
    "    \n",
    "    r_dir = {oligo_list[i]:oligo_freq[i] for i in range(len(oligo_list))}\n",
    "    \n",
    "    for i in range(len(oligo_list)):\n",
    "        oligo_freq[i] += r_dir[rev_comp(oligo_list[i])]\n",
    "    \n",
    "    return oligo_freq\n",
    "\n",
    "\n",
    "def exp_tnf(tetra_list, tri_nucl, di_nucl):\n",
    "    \"\"\"\n",
    "    Calculate expected tetranucleotide frequency\n",
    "    \n",
    "    Inputs:\n",
    "    tetra_list - list of tetranuclotides\n",
    "    tri_nucl - dictionary of trinuclotides frequency\n",
    "    di_nucl - dictionary of dinuclotides frequency\n",
    "    \n",
    "    Output: expected tetranuclotide frequency\n",
    "    \"\"\"\n",
    "    \n",
    "    expected_tnf = np.zeros(len(tetra_list))\n",
    "    for i in range(len(tetra_list)):\n",
    "        tn = tetra_list[i]\n",
    "        expected_tnf[i] = (tri_nucl[tn[0:3]]*tri_nucl[tn[1:4]])/di_nucl[tn[1:3]]\n",
    "            \n",
    "    return expected_tnf\n",
    "\n",
    "\n",
    "def app_variance(expected_tnf, tetra_list, tri_nucl, di_nucl):\n",
    "    \"\"\"\n",
    "    Approximated variance of tetranuclotide frequency\n",
    "    \n",
    "    Inputs:\n",
    "    expected_tnf - expected tetranuclotide frequency\n",
    "    tetra_list - list of tetranuclotides\n",
    "    tri_nucl - dictionary of trinuclotides frequency\n",
    "    di_nucl - dictionary of dinuclotides frequency\n",
    "    \n",
    "    Output - approximated variance of tetranuclotide frequency\n",
    "    \"\"\"\n",
    "    \n",
    "    app_var = np.zeros(len(tetra_list))\n",
    "    for i in range(len(tetra_list)):\n",
    "        tn = tetra_list[i]\n",
    "        n23 = di_nucl[tn[1:3]]\n",
    "        n123 = tri_nucl[tn[0:3]]\n",
    "        n234 = tri_nucl[tn[1:4]]\n",
    "        app_var[i] = expected_tnf[i] * (((n23-n123)*(n23-n234))/(n23*n23))\n",
    "            \n",
    "    return app_var\n",
    "\n",
    "\n",
    "def tnf_z_score(tetra_freq, expected_tnf, app_var):\n",
    "    \"\"\"\n",
    "    Calculate tetranuclotide z-score\n",
    "    \n",
    "    Inputs:\n",
    "    tetra_freq - observed tetranuclotide frequency\n",
    "    expected_tnf - expected tetranuclotide frequency\n",
    "    app_var - approximated variance of tetranuclotide frequency\n",
    "    \n",
    "    Output - tetranuclotide z-score\n",
    "    \"\"\"\n",
    "    \n",
    "    z_scores = (tetra_freq-expected_tnf)/np.sqrt(app_var)\n",
    "    return z_scores\n",
    "\n",
    "\n",
    "def get_z_scores(sequence, di_list, tri_list, tetra_list):\n",
    "    \"\"\"\n",
    "    Calculate z-scores from sequence\n",
    "    \n",
    "    Inputs:\n",
    "    sequence - DNA sequence\n",
    "    di_list - list of possible dinuclotides\n",
    "    tri_list - list of possible trinuclotides\n",
    "    tetra_list - list of possible tetranuclotides\n",
    "    \n",
    "    Output - np array of z-scores\n",
    "    \"\"\"\n",
    "    \n",
    "    tetra_freq, tri_freq, di_freq = get_frequencies(sequence)\n",
    "    tetra_freq = rev_comp_freq(tetra_list, tetra_freq)\n",
    "        \n",
    "    di_nucl = {di_list[i]:di_freq[i] for i in range(len(di_list))}\n",
    "    tri_nucl = {tri_list[i]:tri_freq[i] for i in range(len(tri_list))}\n",
    "    \n",
    "    expected_tnf = exp_tnf(tetra_list, tri_nucl, di_nucl)\n",
    "    app_var = app_variance(expected_tnf, tetra_list, tri_nucl, di_nucl)\n",
    "    z_scores = tnf_z_score(tetra_freq, expected_tnf, app_var)\n",
    "    \n",
    "    return z_scores\n",
    "\n",
    "\n",
    "def inter_correlations(all_z_scores):\n",
    "    \"\"\"\n",
    "    Calcualte Person's correlation coefficient between fragment and median\n",
    "    \n",
    "    Input - np array of z-scores for all fragments\n",
    "    \n",
    "    Output - np array of correlation coeeficients\n",
    "    \"\"\"\n",
    "    \n",
    "    median_z = np.median(all_z_scores, axis=0)\n",
    "\n",
    "    correlations = np.zeros(len(all_z_scores))\n",
    "    for i in range(len(all_z_scores)):\n",
    "        correlations[i] = np.corrcoef(all_z_scores[i], median_z)[1,0]\n",
    "        \n",
    "    return correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tetranucleotide Z-score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kuba/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:110: RuntimeWarning: invalid value encountered in double_scalars\n",
      "/home/kuba/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:134: RuntimeWarning: invalid value encountered in double_scalars\n",
      "/home/kuba/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:151: RuntimeWarning: invalid value encountered in true_divide\n",
      "/home/kuba/anaconda3/lib/python3.7/site-packages/numpy/lib/function_base.py:3405: RuntimeWarning: Invalid value encountered in median for 256 results\n",
      "  r = func(a, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "contigs = SeqIO.parse(\"./test.fasta\", 'fasta')\n",
    "di_list, tri_list, tetra_list = nucl_lists()\n",
    "\n",
    "contigs_z_scores = []\n",
    "for record in contigs:  \n",
    "    z_scores = get_z_scores(record.seq, di_list, tri_list, tetra_list)\n",
    "    contigs_z_scores.append(z_scores)\n",
    "    \n",
    "contigs_z_scores = np.array(contigs_z_scores)\n",
    "\n",
    "contig_corr = inter_correlations(contigs_z_scores)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 0. 0. 0. 0. 2. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      "[0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      " 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n"
     ]
    }
   ],
   "source": [
    "contigs = SeqIO.parse(\"./test.fasta\", 'fasta')\n",
    "di_list, tri_list, tetra_list = nucl_lists()\n",
    "\n",
    "for record in contigs:\n",
    "    tetra_freq, tri_freq, di_freq = get_frequencies(record.seq)\n",
    "    tetra_freq = rev_comp_freq(tetra_list, tetra_freq)\n",
    "\n",
    "    \n",
    "    print(tetra_freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'median_z' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-4f514e58d06d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m plt.fill_between(range(256), np.amax(contigs_z_scores, axis=0), np.amin(contigs_z_scores, axis=0),\n\u001b[1;32m      4\u001b[0m                  alpha=0.4)\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmedian_z\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'median_z' is not defined"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYYAAAD8CAYAAABzTgP2AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAADfJJREFUeJzt221oned9x/GvZhGHtOAn1YljpbOHDcPZoKWSQ1kLWRM/DVKF1i+cvqjoUvxiy4s2BKrgjTZOGHFYcSnNOkxSEIHV7jJKBGGYPDQwykgstxmN1zp23BqrMUmETCALjfF67cV1CZ+/diQf6b6PzjnN9wM3576v+3/O+V+W7d/9pL6UEpIkzfqjTjcgSeouBoMkKTAYJEmBwSBJCgwGSVJgMEiSAoNBkhQYDJKkwGCQJAX9nW5gKd555510/vz5TrchST1laGhoGvjYtep6MhjOnz/P8PBwp9uQpJ6SUmrpiNpLSZKkwGCQJAUGgyQpMBgkSYHBIEkKDAZJUmAwSJICg0GSFBgMkqTAYJAkBQaDJCkwGCRJgcEgSQoMBklSYDBIkgKDQZIUGAySpMBgkCQFBoMkKTAYJEmBwSBJCgwGSVJgMEiSAoNBkhTUFQy7gdPAWWCsyf6VwLGy/2Vg05z9HwfeAx6oqR9J0hLVEQwrgMeBPcA24J7y2uhe4BKwBTgMHJqz/zDw7zX0IkmqqI5g2E4+EzgHXAaOAiNzakaA8bL+NHAH0Fe27y7vPVVDL5KkiuoIho3AhYbtqTI2X80V4F1gHfAR4BvAQzX0IUmqQX8Nn9HXZCy1WPMQ+TLSey18z/6yMDAwsJj+JEmLUEcwTAG3NGwPAm/OUzNVvnMVMAPcBuwFHgNWA78Hfgd8r8n3HCkL09PTc4NHklSTOoLhBLAV2Az8FtgHfGlOzQQwCvwnOQheJJ8xfLah5lvkM4dmoSBJWiZ1BMMV4D7gOPkJpR+QbyQfBCbJofAk8BT5JvUMOTwkSV2oL6XeuyozOTmZhoeHO92GJPWUlNJJYOhadf7msyQpMBgkSYHBIEkKDAZJUmAwSJICg0GSFBgMkqTAYJAkBQaDJCkwGCRJgcEgSQoMBklSYDBIkgKDQZIUGAySpMBgkCQFBoMkKTAYJEmBwSBJCgwGSVJgMEiSAoNBkhQYDJKkwGCQJAUGgyQpMBgkSYHBIEkKDAZJUmAwSJICg0GSFBgMkqSgrmDYDZwGzgJjTfavBI6V/S8Dm8r4DuAk8Ivy+rma+pEkLVEdwbACeBzYA2wD7imvje4FLgFbgMPAoTI+DdwF/DkwCjxVQz+SpArqCIbt5DOBc8Bl4CgwMqdmBBgv608DdwB9wM+BN8v4KeB68tmFJKlD6giGjcCFhu2pMjZfzRXgXWDdnJovkoPigxp6kiQtUX8Nn9HXZCwtsuZW8uWlnQt8z/6yMDAwsJj+JEmLUMcZwxRwS8P2IFcvDzWr6QdWATMN9T8Gvgy8scD3HAGGgKHp6emKLUuS5lNHMJwAtgKbgeuAfcDEnJoJ8s1lgL3Ai+QzhtXAs8CDwE9r6EWSVFEdwXAFuA84DvwS+BH5RvJB4POl5knyPYWzwP1cfaT1PvKTSn8PvFqW9TX0JElaor6U5t4O6H6Tk5NpeHi4021IUk9JKZ0kX5JfkL/5LEkKDAZJUmAwSJICg0GSFBgMkqTAYJAkBQaDJCkwGCRJgcEgSQoMBklSYDBIkgKDQZIUGAySpMBgkCQFBoMkKTAYJEmBwSBJCgwGSVJgMEiSAoNBkhQYDJKkwGCQJAUGgyQpMBgkSYHBIEkKDAZJUmAwSJICg0GSFBgMkqTAYJAkBXUFw27gNHAWGGuyfyVwrOx/GdjUsO/BMn4a2FVTP5KkJaojGFYAjwN7gG3APeW10b3AJWALcBg4VMa3AfuAW8nh8k/l8yRJHVJHMGwnH/GfAy4DR4GROTUjwHhZfxq4A+gr40eBD4Bfl8/ZXkNPkqQlqiMYNgIXGranyth8NVeAd4F1Lb5XkrSM+mv4jL4mY6nFmlbeO2t/WRgYGGi5OUnS4tRxxjAF3NKwPQi8uUBNP7AKmGnxvbOOAEPA0PT0dMWWJUnzqSMYTgBbgc3AdeSbyRNzaiaA0bK+F3iRfGYwUepXlvdvBV6poSdJ0hLVcSnpCnAfcJz8RNEPgFPAQWCS/J//k8BT5JvLM+QwoNT9CPjv8jl/C/xvDT1JkpaoL6X5Lul3r8nJyTQ8PNzpNiSpp6SUTpIvyS/I33yWJAUGgyQpMBgkSYHBIEkKDAZJUmAwSJICg0GSFBgMkqTAYJAkBQaDJCkwGCRJgcEgSQoMBklSYDBIkgKDQZIUGAySpMBgkCQFBoMkKTAYJEmBwSBJCgwGSVJgMEiSAoNBkhQYDJKkwGCQJAUGgyQpMBgkSYHBIEkKDAZJUmAwSJICg0GSFFQNhrXAc8CZ8rpmnrrRUnOmrAPcADwL/Ao4BTxasRdJUg2qBsMY8AKwtbyONalZC3wTuA3YXtZnA+QfgT8FPgn8BbCnYj+SpIqqBsMIMF7Wx4G7m9TsIp9NzACXyvpu4H3gJ6XmMvAzYLBiP5KkiqoGw43AxbJ+EVjfpGYjcKFhe6qMNVoN3EU+65AkdVB/CzXPAzc1GT/Q4nf0NRlLc3r4IfBd4NwCn7O/LAwMDLT41ZKkxWolGO5cYN9bwAby2cIG4O0mNVPA7Q3bg8BLDdtHyDelv3ONPo6Uhenp6XSNWknSElW9lDTB1aeMRoFnmtQcB3aSbzivKevHy75HgFXA1yr2IUmqSdVgeBTYQT7i38HVR06HgCfK+gzwMHCiLAfL2CD5ctQ28o3nV4GvVuxHklRRX0q9d1VmcnIyDQ8Pd7oNSeopKaWT5AP3Bfmbz5KkwGCQJAUGgyQpMBgkSYHBIEkKDAZJUmAwSJICg0GSFBgMkqTAYJAkBQaDJCkwGCRJgcEgSQoMBklSYDBIkgKDQZIUGAySpMBgkCQFBoMkKTAYJEmBwSBJCgwGSVJgMEiSAoNBkhQYDJKkwGCQJAUGgyQpMBgkSYHBIEkKDAZJUlA1GNYCzwFnyuuaeepGS82Zsj7XBPBaxV4kSTWoGgxjwAvA1vI61qRmLfBN4DZge1lvDJAvAO9V7EOSVJOqwTACjJf1ceDuJjW7yGcTM8Clsr677PsocD/wSMU+JEk1qRoMNwIXy/pFYH2Tmo3AhYbtqTIG8DDwbeD9in1IkmrS30LN88BNTcYPtPgdfU3GEvAJYAvwdWBTC5+zvywMDAy0+NWSpMVqJRjuXGDfW8AG8tnCBuDtJjVTwO0N24PAS8CngU8Bvyl9rC/jjbWNjpSF6enp1ELfkqQlqHopaYKrTxmNAs80qTkO7CTfcF5T1o8D3wduJp8tfAZ4nflDQZK0TKoGw6PADvJjqDvKNsAQ8ERZnyHfSzhRloNlTJLUhfpS6r2rMpOTk2l4eLjTbUhST0kpnSQfuC/I33yWJAUGgyQpMBgkSYHBIEkKDAZJUmAwSJICg0GSFBgMkqTAYJAkBQaDJCkwGCRJgcEgSQoMBklSYDBIkgKDQZIUGAySpMBgkCQFBoMkKTAYJEmBwSBJCgwGSVJgMEiSAoNBkhQYDJKkwGCQJAV9KaVO97AU7wDnO93EIg0A051uYpk55w8H59w7/hj42LWKejUYetEkMNTpJpaZc/5wcM5/YLyUJEkKDAZJUmAwLJ8jnW6gA5zzh4Nz/gPjPQZJUuAZgyQpMBjqtRZ4DjhTXtfMUzdaas6U9bkmgNfa0WAbVJnzDcCzwK+AU8Cjbe20ut3AaeAsMNZk/0rgWNn/MrCpYd+DZfw0sKutXdZrqXPeAZwEflFeP9fuRmtS5WcM8HHgPeCB9rW4DFJKLvUtj6WUxsr6WErpUJOatSmlc+V1TVlf07D/Cymlf0kpvdYF82n3nG9IKf1lqbkupfQfKaU9XTCnZsuKlNIbKaU/Kb3+V0pp25yav0kp/XNZ35dSOlbWt5X6lSmlzeVzVnTBnNo550+mlG4u63+WUvptF8ynnfOdXf4tpfSvKaUHumA+S148Y6jXCDBe1seBu5vU7CIfWc8Al8r67rLvo8D9wCPtbbNWVeb8PvCTUnMZ+Bkw2M5mK9hOPko8R+71KHnujRr/LJ4G7gD6yvhR4APg1+Vztre/5cqqzPnnwJtl/BRwPflou5tVmS/kv/vnyPPtaQZDvW4ELpb1i8D6JjUbgQsN21NlDOBh4Nvk/zB7RdU5z1oN3AW8UHeDNWllDo01V4B3gXUtvrcbVZlzoy+Sg+KDNvRYpyrz/QjwDeChNve4LPo73UAPeh64qcn4gRbf39dkLAGfALYAX+f/X7fstHbNeVY/8EPgu+Qjrm50rTksVNPKe7tRlTnPuhU4BOysq6k2qjLfh4DD5PsLPc9gWLw7F9j3FrCBfOS8AXi7Sc0UcHvD9iDwEvBp4FPAb8g/l/VlvLG2U9o151lHyDelv1OlyTabAm5p2B7k6qWSuTVT5J/hKvLls1be242qzHm2/sfAl4E32tppParM9zZgL/AY+ez398DvgO+1t+X28FJSvSa4+sTNKPBMk5rj5KOnNWXZWca+D9xMPlv4DPA63REK11JlzpDvp6wCvtbeNis7AWwFNgPXAfvIc2/U+GexF3iRfDQ5UepXlvdvBV5pf8uVVZnzavITZw8CP12OZmtQZb6fJf/b3UQ+wPkHejQUAJ9KqnlZl1J6IaV0pryuLeNDKaUnGur+OqV0tixfafI5m1LvPJVUZc6DKftlSunVsny1C+Y03/JXKaXXU35y5UAZO5hS+nxZvz7lJ1LOppReSfnpltn3HijvO52698mrOuf8dyml/2n4ub6aUlrfBfNp5894dvlW6vGnkvzNZ0lS4KUkSVJgMEiSAoNBkhQYDJKkwGCQJAUGgyQpMBgkSYHBIEkK/g81G71kxY3e7gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "dark"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plt.plot(np.amax(all_z_scores, axis=0), alpha=0.4)\n",
    "#plt.plot(np.amin(all_z_scores, axis=0), alpha=0.4)\n",
    "plt.fill_between(range(256), np.amax(contigs_z_scores, axis=0), np.amin(contigs_z_scores, axis=0),\n",
    "                 alpha=0.4)\n",
    "plt.plot(median_z)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
